usr/src/uts/common/fs/zfs/dmu_send.c

```diff
@@ -2117,27 +2118,47 @@ receive_object(struct receive_writer_arg *rwa, struct drr_object *drro,
 
 	tx = dmu_tx_create(rwa->os);
 	dmu_tx_hold_bonus(tx, object);
+	dmu_tx_hold_write(tx, object, 0, 0);
 	err = dmu_tx_assign(tx, TXG_WAIT);
 	if (err != 0) {
 		dmu_tx_abort(tx);
 		return (err);
 	}
 
+	db = NULL;
 	if (object == DMU_NEW_OBJECT) {
 		/* currently free, want to be allocated */
 		err = dmu_object_claim(rwa->os, drro->drr_object,
 		    drro->drr_type, drro->drr_blksz,
 		    drro->drr_bonustype, drro->drr_bonuslen, tx);
 	} else if (drro->drr_type != doi.doi_type ||
-	    drro->drr_blksz != doi.doi_data_block_size ||
-	    drro->drr_bonustype != doi.doi_bonus_type ||
-	    drro->drr_bonuslen != doi.doi_bonus_size) {
+	    (drro->drr_blksz != doi.doi_data_block_size &&
+	     doi.doi_max_offset > doi.doi_data_block_size)) {
```

Comment:
The case of type change I found is zfs_sa_upgrade().  But as I understand (may be wrong) before that spill blocks were not really used, so it should not matter.  On the other side, I am not sure about general policy -- does spill block content formally depend on bonus buffer type and should be dropped and rewritten in case of any change, or there can be cases (non-SA) where they are independent?

Reply:
The spill block is an extension of the bonus buffer and should always have the same type as it. I cannot think of any cases where the bonus type is changed without re-creating the entire object. I could be wrong on this though.

Reply:
The DMU supports changing the bonus type and length of an existing object.  I'm not sure if the ZPL actually uses that functionality when also using spill blocks, so we may be safe, but this still seems fragile.  If this code is going to assume that changed bonus type == spill block doesn't need to be preserved, we should also add checks when changing the bonus type that the spill block does not exist.